### DISCUSSION TITLE
AssetSpec provides no value

### DISCUSSION THREAD

**Discussion by mattfysh** [dagster-io/dagster#23783] [Q&A]
AssetSpec is intended to replace SourceAsset, but you can no longer attach an IOManager to it, so a Basic Asset is required to implement a central loading mechanism (without repeating yourself in downstream assets).

Using the example from the docs on [External Assets](https://docs.dagster.io/concepts/assets/external-assets)

```py
from dagster import AssetSpec, Definitions, asset

raw_logs = AssetSpec("raw_logs")
processed_logs = AssetSpec("processed_logs", deps=[raw_logs])


@asset(deps=[processed_logs])
def aggregated_logs() -> None:
    # Loads "processed_log" into memory and performs some aggregation
    ...
```

This is very poor design, what if you have multiple downstream assets that depend on `processed_logs`? You would want to centralize the loading architecture, so you are now required to create a new Basic Asset to do so

```py
processed_logs = AssetSpec("processed_logs", deps=[raw_logs])

@asset(deps=[processed_logs])
def loaded_logs() -> None:
    # Loads "processed_log"
    ...

@asset(deps=[loaded_logs])
def aggregated_logs() -> None:
    ...

@asset(deps=[loaded_logs])
def analyzed_logs() -> None:
    ...
```

What is the value in keeping `processed_logs` around any longer as an AssetSpec? The logical & physical asset is already represented now by `loaded_logs`

**Comment by zyd14** [comment]
I think saying AssetSpec provides no value is a bit hyperbolic, but I do agree that eliminating SourceAssets without an equivalent to replace it is difficult to see. My team uses SourceAssets quite a bit, and we really like the ability to attach an IO manager to them so we can treat them like any other asset. I don't really understand why this feature is being removed entirely, couldn't it just be made optional through the `deps` mechanism? What is the benefit of external assets over SourceAssets? Why does every consumer of an external asset have to define how it loads the external asset? Is it really that common to load a source in a bunch of different ways? If so, don't input loaders help with that?

It just sucks to see that a component you rely on and were advised to use by the Dagster team in discussions as recently as April is now being deprecated. It wouldn't be that big of a deal if there was an equivalent being offered, but that doesn't seem to be the case. So now it seems like the only option is to go edit a few dozen asset definitions to insert input-loading logic in addition to the IO manager logic they already have for loading other non-source assets. 

To put it more generally, what happens when you have an asset that has a normal asset and an external asset as upstream dependencies? Do you have to define input-loading logic in the asset definition AND in the IO manager? Or are you expected to just drop IO managers entirely and stick all the input-loading logic into your asset definition?

**Reply by mattfysh** [reply]
The awkwardness comes from defining input-loading in an unrelated asset, instead of placing it where you're defining the upstream asset.

For now I've skipped AssetSpec completely and replaced my SourceAssets with basic `@asset` definitions. I'm not sure really how much of a difference it makes, other than being able to materialize something you shouldn't be able to - so in the `handle_output` method of the IO manager I just raise an exception.
